File: samples/JavaScript/chat-server.js
var net = require("net");
var sys = require("sys");
var EventEmitter = require("events").EventEmitter;

/*******************************************************************************
 * ChatServer
 *
 * Manages connections, users, and chat messages.
 ******************************************************************************/

function ChatServer() {
  this.chatters = {};
  this.server   = net.createServer(this.handleConnection.bind(this));
  this.server.listen(1212, "localhost");
}

ChatServer.prototype.isNicknameLegal = function(nickname) {
  // A nickname may contain letters or numbers only,
  // and may only be used once.
  if(nickname.replace(/[A-Za-z0-9]*/, '') != "") {
    return false
  }
  for(used_nick in this.chatters) {
    if(used_nick == nickname) {
      return false;
    }
  }
  return true;
};

ChatServer.prototype.handleConnection = function(connection) {
  console.log("Incoming connection from " + connection.remoteAddress);
  connection.setEncoding("utf8");

  var chatter = new Chatter(connection, this);
  chatter.on("chat", this.handleChat.bind(this));
  chatter.on("join", this.handleJoin.bind(this));
  chatter.on("leave", this.handleLeave.bind(this));
};

ChatServer.prototype.handleChat = function(chatter, message) {
  this.sendToEveryChatterExcept(chatter, chatter.nickname + ": " + message);
};

ChatServer.prototype.handleJoin = function(chatter) {
  console.log(chatter.nickname + " has joined the chat.");
  this.sendToEveryChatter(chatter.nickname + " has joined the chat.");
  this.addChatter(chatter);
};

ChatServer.prototype.handleLeave = function(chatter) {
  console.log(chatter.nickname + " has left the chat.");
  this.removeChatter(chatter);
  this.sendToEveryChatter(chatter.nickname + " has left the chat.");
};

ChatServer.prototype.addChatter = function(chatter) {
  this.chatters[chatter.nickname] = chatter;
};

ChatServer.prototype.removeChatter = function(chatter) {
  delete this.chatters[chatter.nickname];
};

ChatServer.prototype.sendToEveryChatter = function(data) {
  for(nickname in this.chatters) {
    this.chatters[nickname].send(data);
  }
};

ChatServer.prototype.sendToEveryChatterExcept = function(chatter, data) {
  for(nickname in this.chatters) {
    if(nickname != chatter.nickname) {
      this.chatters[nickname].send(data);
    }
  }
};

/*******************************************************************************
 * Chatter
 *
 * Represents a single user/connection in the chat server.
 ******************************************************************************/

function Chatter(socket, server) {
  EventEmitter.call(this);

  this.socket     = socket;
  this.server     = server;
  this.nickname   = "";
  this.lineBuffer = new SocketLineBuffer(socket);

  this.lineBuffer.on("line", this.handleNickname.bind(this));
  this.socket.on("close", this.handleDisconnect.bind(this));

  this.send("Welcome! What is your nickname?");
};

sys.inherits(Chatter, EventEmitter);

Chatter.prototype.handleNickname = function(nickname) {
  if(server.isNicknameLegal(nickname)) {
    this.nickname = nickname;
    this.lineBuffer.removeAllListeners("line");
    this.lineBuffer.on("line", this.handleChat.bind(this));
    this.send("Welcome to the chat, " + nickname + "!");
    this.emit("join", this);
  } else {
    this.send("Sorry, but that nickname is not legal or is already in use!");
    this.send("What is your nickname?");
  }
};

Chatter.prototype.handleChat = function(line) {
  this.emit("chat", this, line);
};

Chatter.prototype.handleDisconnect = function() {
  this.emit("leave", this);
};

Chatter.prototype.send = function(data) {
  this.socket.write(data + "\r\n");
};

/*******************************************************************************
 * SocketLineBuffer
 *
 * Listens for and buffers incoming data on a socket and emits a 'line' event
 * whenever a complete line is detected.
 ******************************************************************************/

function SocketLineBuffer(socket) {
  EventEmitter.call(this);

  this.socket = socket;
  this.buffer = "";

  this.socket.on("data", this.handleData.bind(this));
};

sys.inherits(SocketLineBuffer, EventEmitter);

SocketLineBuffer.prototype.handleData = function(data) {
  for(var i = 0; i < data.length; i++) {
    var char = data.charAt(i);
    this.buffer += char;
    if(char == "\n") {
      this.buffer = this.buffer.replace("\r\n", "");
      this.buffer = this.buffer.replace("\n", "");
      this.emit("line", this.buffer);
      this.buffer = "";
    }
  }
};

// Start the server!
server = new ChatServer();

File: samples/Erlang/chat-server.erl
-module(chat).

-export([start/0, start/1]).

-record(client, {name=none, socket=none}).

start() -> start(8080).
start(Port) ->
    register(server, spawn(fun() -> server() end)),
    {ok, LSocket} = gen_tcp:listen(Port, [binary, {packet, 0}, {active, false}, {reuseaddr, true}]),
    accept(LSocket).

% main loop for message dispatcher
server() -> server([]).
server(Clients) ->
    receive
        {join, Client=#client{name = Name, socket = Socket}} ->
            self() ! {say, Socket, "has joined." ++ [10, 13]},
            server(Clients ++ [Client]);
        {leave, Socket} ->
            {value, #client{name = Name}, List} = lists:keytake(Socket, 3, Clients),
            self() ! {say, none, Message = "has left."},
            server(List);
        {say, Socket, Data} ->
            {value, #client{name = From}, List} = lists:keytake(Socket, 3, Clients),
            Message = From ++ " : " ++ Data,
            lists:map(fun(#client{socket = S}) ->
                    gen_tcp:send(S, Message)
                end, List)
    end,
    server(Clients).

% accepts connections then spawns the client handler
accept(LSocket) ->
    {ok, Socket} = gen_tcp:accept(LSocket),
    spawn(fun() -> connecting(Socket) end),
    accept(LSocket).

% when client is first connect send prompt for user name
connecting(Socket) ->
    gen_tcp:send(Socket, "What is your name? "),
    case listen(Socket) of
        {ok, N} ->
            Name = binary_to_list(N),
            server ! {join, #client{name =  lists:sublist(Name, 1, length(Name) - 2), socket = Socket} },
            client(Socket);
        _ -> ok
    end.

% main client loop that listens for data
client(Socket) ->
    case listen(Socket) of
        {ok, Data} ->
            server ! {say, Socket, binary_to_list(Data)},
            client(Socket);
        _ -> server ! {leave, Socket}
    end.

% utility function that listens for data on a socket
listen(Socket) ->
    case gen_tcp:recv(Socket, 0) of
        Response -> Response
    end.
